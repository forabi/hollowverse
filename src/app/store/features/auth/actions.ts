import { createActionCreator } from 'store/helpers';

export const requestLogin = createActionCreator('REQUEST_LOGIN');
export const requestLogout = createActionCreator('REQUEST_LOGOUT');

export const setFbSdkAuthState = createActionCreator('SET_FB_SDK_AUTH_STATE');

export const facebookAuthResponseChanged = createActionCreator(
  'FACEBOOK_AUTH_RESPONSE_CHANGED',
);
