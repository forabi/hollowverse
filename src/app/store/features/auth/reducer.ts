import { isActionOfType } from 'store/helpers';
import { StoreState, AuthState, Reducer } from 'store/types';
import { createSelector } from 'reselect';
import { getResolvedDataForKey } from '../asyncData/selectors';
import { isPendingResult, isSuccessResult } from 'helpers/asyncResults';

export const fbSdkAuthStateReducer: Reducer<StoreState['fbSdkAuthState']> = (
  state,
  action,
) => {
  if (isActionOfType(action, 'SET_FB_SDK_AUTH_STATE')) {
    return action.payload;
  }

  if (isActionOfType(action, 'FACEBOOK_AUTH_RESPONSE_CHANGED')) {
    if (action.payload) {
      const { accessToken } = action.payload;

      return { state: 'loggedIn', accessToken };
    }

    return { state: 'loggedOut' };
  }

  if (isActionOfType(action, 'REQUEST_LOGOUT')) {
    return { state: 'loggingOut' };
  }

  return (
    state || {
      state: 'initializing',
    }
  );
};

export const getFbSdkAuthState = (state: StoreState) => state.fbSdkAuthState;

export const getAccessToken = createSelector(
  getFbSdkAuthState,
  fbAuthState =>
    fbAuthState.state === 'loggedIn' ? fbAuthState.accessToken : null,
);

export const getViewerResult = createSelector(getResolvedDataForKey, get =>
  get('viewer'),
);

export const getAuthState = createSelector(
  getViewerResult,
  getFbSdkAuthState,
  (viewerQueryResult, fbAuthState): AuthState => {
    if (fbAuthState.state !== 'loggedIn') {
      return fbAuthState;
    }

    if (isPendingResult(viewerQueryResult)) {
      return { state: 'loggingIn' };
    }

    if (
      isSuccessResult(viewerQueryResult) &&
      viewerQueryResult.value.viewer !== null
    ) {
      return { state: 'loggedIn', viewer: viewerQueryResult.value.viewer };
    }

    return { state: 'error' };
  },
);
