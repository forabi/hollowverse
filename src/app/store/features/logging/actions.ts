import { createActionCreator } from 'store/helpers';

export const unhanldedErrorThrown = createActionCreator(
  'UNHANDLED_ERROR_THROWN',
);

export const urlClicked = createActionCreator('URL_CLICKED');

export const pageRenderStarted = createActionCreator('PAGE_RENDER_STARTED');
export const pageLoadSucceeded = createActionCreator('PAGE_LOAD_SUCCEEDED');
export const pageLoadFailed = createActionCreator('PAGE_LOAD_FAILED');
