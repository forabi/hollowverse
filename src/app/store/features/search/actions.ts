import { createActionCreator } from 'store/helpers';

export const requestSearchResults = createActionCreator(
  'REQUEST_SEARCH_RESULTS',
);
export const setSearchResults = createActionCreator('SET_SEARCH_RESULTS');
export const setSearchError = createActionCreator('SET_SEARCH_ERROR');
