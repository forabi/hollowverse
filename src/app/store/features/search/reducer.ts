import { createReducerForStoreKey, handleAction } from 'store/helpers';

export const searchResultsReducer = createReducerForStoreKey<'searchResults'>(
  {
    SET_SEARCH_RESULTS: handleAction('SET_SEARCH_RESULTS'),
    REQUEST_SEARCH_RESULTS: handleAction('REQUEST_SEARCH_RESULTS'),
  },
  {
    hasError: false,
    isInProgress: false,
    value: null,
  },
);

export const lastSearchMatchReducer = createReducerForStoreKey<
  'lastSearchMatch'
>(
  {
    SET_LAST_SEARCH_MATCH: handleAction('SET_LAST_SEARCH_MATCH'),
  },
  null,
);
