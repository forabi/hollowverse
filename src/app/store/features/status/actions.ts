import { createActionCreator } from 'store/helpers';

export const setStatusCode = createActionCreator('SET_STATUS_CODE');
export const setRedirectionUrl = createActionCreator('SET_REDIRECTION_URL');
