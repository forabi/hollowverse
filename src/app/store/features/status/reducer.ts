import { createReducerForStoreKey, handleAction } from 'store/helpers';
import { StoreState } from 'store/types';

export const statusCodeReducer = createReducerForStoreKey<'statusCode'>(
  {
    SET_STATUS_CODE: handleAction('SET_STATUS_CODE'),
  },
  200,
);

export const getStatusCode = (state: StoreState) => state.statusCode;
