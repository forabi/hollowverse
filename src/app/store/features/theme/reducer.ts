import { createReducerForStoreKey, isActionOfType } from 'store/helpers';
import { StoreState } from 'store/types';
import { createSelector } from 'reselect';
import { createMuiTheme } from '@material-ui/core/styles';
import green from '@material-ui/core/colors/green';
import orange from '@material-ui/core/colors/orange';
import indigo from '@material-ui/core/colors/indigo';
import yellow from '@material-ui/core/colors/yellow';

export const themeReducer = createReducerForStoreKey<'theme'>(
  {
    TOGGLE_NIGHT_MODE: (state, action) => {
      if (isActionOfType(action, 'TOGGLE_NIGHT_MODE')) {
        return state === 'light' ? 'dark' : 'light';
      }

      return state;
    },
  },
  'light',
);

export const getTheme = (state: StoreState) => state.theme;

export const isNightModeEnabled = createSelector(
  getTheme,
  theme => theme === 'dark',
);

export const getMuiTheme = createSelector(getTheme, theme =>
  createMuiTheme({
    breakpoints: {
      values: {
        xs: 0,
        sm: 720,
        md: 960,
        lg: 1280,
        xl: 1920,
      },
    },
    typography: {
      fontFamily: 'var(--font-family)',
      fontSize: 14,
      display1: {
        fontSize: 26,
      },
      body1: {
        fontSize: 16,
        lineHeight: 1.6,
      },
      body2: {
        fontSize: 15,
      },
      title: {
        lineHeight: 1.6,
        fontWeight: 700,
      },
    },
    palette: {
      type: theme,
      primary: theme === 'dark' ? yellow : indigo,
      secondary: theme === 'dark' ? orange : green,
    },
    overrides: {
      MuiTypography: {
        gutterBottom: {
          marginBottom: '15px',
        },
      },
      MuiMenu: {
        paper: {
          minWidth: 250,
        },
      },
      MuiPaper: {
        root: {
          marginBottom: 10,
        },
      },
      MuiTooltip: {
        tooltip: {
          fontSize: 13,
        },
      },
      MuiExpansionPanel: {
        expanded: {
          marginTop: 0,
          '&::before': {
            opacity: 'inherit',
          },
        },
      },
    },
    props: {
      MuiTooltip: {
        enterDelay: 400,
      },
      MuiPaper: {
        elevation: 1,
      },
    },
  }),
);
