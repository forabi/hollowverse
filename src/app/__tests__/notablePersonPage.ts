import { getStatusCode } from 'store/features/status/reducer';
import {
  createMockGetResponseForDataRequest,
  TestContext,
  createTestContext,
} from 'helpers/testHelpers';
import { pageLoadSucceeded } from 'store/features/logging/actions';

describe('Notable Person page', () => {
  let context: TestContext;

  describe('When notable person is found,', () => {
    beforeEach(async () => {
      context = await createTestContext({
        epicDependenciesOverrides: {
          getResponseForDataRequest: createMockGetResponseForDataRequest(
            'notablePersonQuery',
            {
              notablePerson: {
                commentsUrl: '',
                name: 'Tom Hanks',
                editorialSummary: null,
                mainPhoto: null,
                relatedPeople: [
                  {
                    mainPhoto: null,
                    name: 'Al Pacino',
                    slug: 'Al_Pacino',
                  },
                ],
                slug: 'Tom_Hanks',
                summary: null,
              },
            },
          ),
        },
        createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
      });
    });

    it('returns 200', () => {
      expect(getStatusCode(context.store.getState())).toBe(200);
    });

    it('has notable person name', () => {
      expect(context.wrapper).toIncludeText('Tom Hanks');
    });

    it('shows related people', () => {
      expect(context.wrapper).toIncludeText('Al Pacino');
    });

    describe('logs page load event', () => {
      beforeEach(done => {
        window.addEventListener('unload', () => {
          done();
        });

        window.dispatchEvent(new Event('unload'));
      });

      it('sends logs on page unload', () => {
        expect(context.dependencies.sendLogs).toHaveBeenLastCalledWith(
          expect.arrayContaining([
            pageLoadSucceeded(
              context.history.createHref(context.history.location),
            ),
          ]),
        );
      });
    });
  });

  describe('When notable person is not found,', () => {
    beforeEach(async () => {
      context = await createTestContext({
        createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
        epicDependenciesOverrides: {
          getResponseForDataRequest: createMockGetResponseForDataRequest(
            'notablePersonQuery',
            {
              notablePerson: null,
            },
          ),
        },
      });
    });

    it('returns 404', () => {
      expect(getStatusCode(context.store.getState())).toBe(404);
    });

    it('shows "Not Found"', () => {
      expect(context.wrapper).toIncludeText('Not Found');
    });
  });
});
