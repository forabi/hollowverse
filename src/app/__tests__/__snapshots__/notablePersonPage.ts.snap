// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notable person page when notable person is found, if notable person has an editorial summary shows editorial summary content 1`] = `
<article
  class="PureComponent-article"
>
  <div
    aria-hidden="false"
    class="PureComponent-root"
  >
    <div
      aria-hidden="true"
      class="PureComponent-coverBackgroundWrapper"
    >
      <div
        class="PureComponent-coverBackground"
      />
    </div>
    <div
      class="PureComponent-container"
    >
      <span
        class="PureComponent-photoLink"
      >
        <img
          class="PureComponent-root PureComponent-photo"
          role="presentation"
          src="[object Object]"
        />
      </span>
      <div
        class="PureComponent-content"
      >
        <h1
          class="MuiTypography-root MuiTypography-display1 MuiTypography-gutterBottom MuiTypography-alignCenter"
        >
          <small
            class="MuiTypography-root MuiTypography-caption MuiTypography-alignCenter"
            style="text-transform: uppercase; letter-spacing: 1px; font-size: 11px;"
          >
            <span
              class="PureComponent-text"
            >
              Religion, politics, and ideas of
            </span>
          </small>
          <span
            class="PureComponent-text PureComponent-name"
          >
            Tom Hanks
          </span>
        </h1>
      </div>
    </div>
  </div>
  <div
    class="PureComponent-stub"
  >
    Share what you know about the religion and political views of 
    Tom Hanks
     
    in the comments below
  </div>
</article>
`;
